File: csharp-solutions/Arrays/Sliding Window/76.MinimumWindowSubstring.cs
/** 
Link: https://leetcode.com/problems/minimum-window-substring/
Author: Nicolas Taveira
Date: 16/02/21
**/

namespace MinimumWindowSubstring{
   public class Solution {
    public string MinWindow(string s, string t) {
        var counterT = new Dictionary<char, int>();
        
        foreach(var curr in t)
        {
            if(counterT.ContainsKey(curr)){
                counterT[curr]++;
            }else{
                counterT.Add(curr, 1);
            }
        }
        
        var filteredS = new List<(char c, int pos)>();
        for(var index = 0; index < s.Length; index++){
            var curr = s[index];
            if(counterT.ContainsKey(curr)){
                filteredS.Add((curr, index));
            }
        }
        
        var left = 0;
        var answer = (size: s.Length + 1, start: 0, end: 0);
        var counterS = new Dictionary<char,int>();
        var matches = 0;
        
        for(int right= 0; right < filteredS.Count; right++){
            (char c, int pos) = filteredS[right];
            
            int currCount;
            if(counterS.ContainsKey(c)){
                counterS[c]++;
                currCount = counterS[c];
            }else{
                currCount = 1;
                counterS.Add(c, currCount);
            }
            
            if(currCount == counterT[c]) matches++;
            
            while(left <= right && matches == counterT.Count){
                var firstChar = filteredS[left];
                var currSize = pos - firstChar.pos + 1;
                
                if(currSize <= answer.size){
                    answer = (currSize, firstChar.pos, pos);
                }
                
                
                if(counterS[firstChar.c] == counterT[firstChar.c]) matches--;
                counterS[firstChar.c]--;
                left++;
                
            }
            
        }
        
        if(answer.size > s.Length) return "";
        
        return s.Substring(answer.start, answer.size);
    }
}
}

File: python-solutions/Arrays/Sliding Window/76.MinimumWindowSubstring.py
'''
Link: https://leetcode.com/problems/minimum-window-substring/
Author: Nicolas Taveira
Date: 16/02/21
'''

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        counter_t = {}
        
        for char in t:
            counter_t[char] = counter_t.get(char, 0) + 1
        
        
        filtered_s = []
        for index, char in enumerate(s):
            if char in counter_t:
                filtered_s.append((index, char))
                
        
        required_matches = len(counter_t)
        left, char_matches = 0, 0
        counter_s = {}
        
        answer = len(s), - 1, -1
 
        for right in range(len(filtered_s)):
            s_index, curr_char = filtered_s[right]
            
            counter_s[curr_char] = counter_s.get(curr_char, 0) + 1
            
            if counter_s[curr_char] == counter_t[curr_char]: 
                char_matches += 1
            
            while left <= right and char_matches == required_matches:
                curr_size = s_index - filtered_s[left][0] + 1
                
                if answer[0] >= curr_size:
                    answer = curr_size, filtered_s[left][0], s_index
                
                
                first_char = filtered_s[left][1]
                
                if counter_s[first_char] == counter_t[first_char]:
                    char_matches -= 1
                
                counter_s[first_char] -= 1
                left += 1
                
        return s[answer[1]: answer[2] + 1]